Add async support to `State.result()` (#7071)
PrefectHQ/prefect: docs/concepts/results.md

my_flow()
```

### Working with async results

When **calling** flows or tasks, the result is returned directly:

```python
import asyncio
from prefect import flow, task

@task
async def my_task():
    return 1

@flow
async def my_flow():
    task_result = await my_task()
    return task_result + 1

result = asyncio.run(my_flow())
assert result == 2
```

When working with flow and task states, the result can be retreived with the `State.result()` method:

```python
import asyncio
from prefect import flow, task

@task
async def my_task():
    return 1

@flow
async def my_flow():
    state = await my_task(return_state=True)
    result = await state.result(fetch=True)
    return result + 1

async def main():
    state = await my_flow(return_state=True)
    assert await state.result(fetch=True) == 2

asyncio.run(main())
```

!!! important "Resolving results"

>     In Prefect 2.6.0, we added automatic retrieval of persisted results.

Reviewer: ```suggestion
    Prefect 2.6.0 added automatic retrieval of persisted results.
```

    Prior to this version, `State.result()` did not require an `await`.

>     For backwards compatibility, when used from an asynchronous context, `State.result()` will return a raw result type.

Reviewer: ```suggestion
    For backwards compatibility, when used from an asynchronous context, `State.result()` returns a raw result type.
```

>     You may opt-in to the new behavior by passing `fetch=True` as shown in the example above.

Reviewer 2: ```suggestion
    
    You may opt-in to the new behavior by passing `fetch=True` as shown in the example above.
```

Reviewer 2: Break up concepts into some paragraphs for easier scanning.

    If you would like this behavior to be used automatically, you may enable the `PREFECT_ASYNC_FETCH_STATE_RESULT` setting.

>     If you do not opt-in to this behavior you will see a warning.

Reviewer: ```suggestion
    If you do not opt-in to this behavior, you will see a warning.
```

>     You may also opt-out by setting `fetch=False`.

Reviewer 2: ```suggestion
    
    You may also opt-out by setting `fetch=False`.
```

>     This will silence the warning but you will need to retrieve your result manually from the result type.

Reviewer 2: ```suggestion
    This will silence the warning, but you will need to retrieve your result manually from the result type.
```
